### ISSUE TITLE
new API

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/dataflower#119]
I think the current API with push and pull and pump and flows is way too complicated. I'd like something different and something simple. Maybe starting from flow abstraction level was a mistake and I should start from high level.

I thought of something like this:

```js
fn = df(fn1).await(fn2)
fn()
```

which means

```js
fn = function (){
    args = a(arguments)
    cb = function (){
        fn1.apply(this, arguments);
    }
    args.unshift(cb)
    fn2.apply(this, args)
}
fn()
```

**Comment by inf3rno** [comment]
I need to reinforce, that this is not the proper way of treating async code and using async components, ports, etc what I really need.

**Comment by inf3rno** [comment]
Actually this is just an API, it does not show the underlying structure, so it could work with the current implementation as well. I worked with the current code 4 months ago, I barely remember why I chose this implementation. :S

**Comment by inf3rno** [comment]
What is the main problem by async programming?

The callback hell
```js
a(function (x){
     b(function (y){
         c(function (z){
              ....
         }, y)
     }, x)
})
```

How to flat this structure?

```js
cb1 = function (x){
    b(cb2, x)
}
cb2 = function (y){
    c(cb3, y)
}
cb3 = function (z){
    ....
}
a(cb1)
```

What's the problem with this?

It is hard to follow which step comes after the other. By sync code we would do something like this:

```js
x = a()
y = b(x)
z = c(y)
...
```

By ES7 async functions we can do more or less the same.

How to do something similar by ES5?

```js
df(a)
    .pipe(b)
    .pipe(c)
    .pipe(...)
    .call()
```

Where pipe automatically generates the callbacks which call the functions and pass the arguments.

This is what the async lib does, and I am sure I don't want another async lib, since it is already implemented.

So let's talk about what we can add to this, which worth creating another lib.

**Comment by inf3rno** [comment]
What we have here is a call graph https://en.wikipedia.org/wiki/Call_graph . 

```
a -(x)-> b -(y)-> c -(z)->
```

we can describe that graph with nodes and links

```js
A = new node(a)
B = new node(b)
C = new node(c)
AB = new link(A, B)
BC = new link(B, C)
...
A.call()
```

Okay this is not something we want to read in the code, but we can make a graph building API which is simpler:

```js
node(a).link(b).link(c).call()
// node - creating a new node
// link - creating a new node, and linking it from the last added node
// call - calls the first added node and so the whole graph
```

which is almost the same as the async lib.

Using nodes have the advantage that we can always remove a node or add a new one. This means the existing call graph is totally reusable and can be changed on the fly.

We can add a new node to the graph:

```
a -(x)-> b -(y)-> c -(z)->
=>
a -(x)-> b -(y)-> q(z) -> c -(p)->
```

```js
cb1 = function (x){
    b(cb2, x)
}
cb2 = function (y){
    c(cb3, y)
}
cb3 = function (z){
    ....
}
a(cb1)

cbx = cb3
cb3 = function (){
    d(cbx)
}
a(cb1)
```

or

```js
remove(BC)
add(Q)
add(BQ)
add(QC)
```

Note that the async lib does not support this feature

This way we can write self modifying code and probably we can modify the running code externally by adding new nodes and links to it. It is possible but that is not an advantage we can exploit in an average code.

**Comment by inf3rno** [comment]
Okay, let's talk about another aspect: pull/push.

Pull and push are interesting terms. Pull is the analogy of return/callback and push is the analogy of call. By push we send variables to a component and by pull we receive variables from a component. By push we decide when to send the variables, since we have them. By pull the sender decides when to send us the variables, but we can trigger the variable generating process by sending a pull notification.

push:
```js
a(x) // (a)sync
```

pull:
```js
y = b() // sync
b(function (y){}) // async
```

We can describe the call graph:

```
a -(x)-> b -(y)-> c -(z)->
```

```js
-push-> a -push(x)-> b -push(y)-> c -push(z)->
```

But we can describe the call graph in a different way too:

```js
<-pull(z)- c <-pull(y)- b <-pull(x)- a
```

Why is the important?

This is because async pull is supported by async functions only. So in order to fully support pull we need to add a callback to every function. By node.js this is not a problem, by client side js this is not that usual.

Why is pull better than push?

It is really simple, that's why we use callbacks. By push the function needs to know where to send the results (strong coupling), by pull the functions does not need to know where to send the results (loose coupling). Is this really that loose?

Let's check the code:

```js
cb1 = function (x){
    b(cb2, x)
}
cb2 = function (y){
    c(cb3, y)
}
cb3 = function (z){
    ....
}
a(cb1)
```

a does not know cb1
cb1 knows about b and cb2
b does not know cb2
cb2 knows about c and cb3
c does not know cb3

So the callback definitions contain the strong coupled part. Now that part must go to a higher abstraction level, since that describes the call graph.

**Comment by inf3rno** [comment]
Is there a way to combine the pull/push logic with the node/link logic? How does the node/link logic work?

```js
a = node(function (done, x){
    done(y)
})
b = node(function(done, y){
    done()
})
ab = link(a, b) // a->b
a.call(null, x)
// a(x) -> y, b(y) -> -
```

```js
node = function (fn){
    links = [] 
    wrapper = function (){
        done = function (){
            links.forEach(function (link){
                link.target.apply(null, arguments)
            })
        }
        args = array(arguments)
        args.unshift(done)
        fn.apply(this, arguments)
    }
    wrapper.links = links
    return wrapper
}

link = function (source, target){
    link = {source: source, target: target}
    source.links.push(link)
    return link
}
```

This is the implementation of the push part (`a` pushes to `b`).

How to implement the pull part?

**Comment by inf3rno** [comment]
What is the pull about?

The pull means that `y` is a requirement of `b` and to call `b` you need to pull in that requirement from `a`. Let's say that `a` don't have a param, or `b` knows the param if it has. That would mean:

```js
function a(x){
    return y
}
function b(){
    y = a(x)
}
```

but this is not entirely true. Something like this would be nice if we want to pull in arguments:

```js
function a(x){
    return y
}
function b(y = a(x)){
}
```

This is possible if we don't call `b` before getting the results of `a`. How would this look like with the node logic?

```js
a = node(function (done, x){
    done(y)
})
b = node(function (done, y){
    done()
})
link(a, b)
b()
```

Now before calling `b` the wrapper calls `a` to get the `y` argument. The only problem, that `x` is unknown in this system, so `a` cannot run. This can be solved by:

```js
link(a, b, [x])
```

so if `a` is pulled, then `x` is given by the link.

So

```js
df(a).link(b, [x])
```

can be a good solution, if we are not sure whether `b` will be called directly, or through `a`. By direct call `b` will pull in args from `a` otherwise `a` will push args to `b`.

This is an interesting system, I'll develop this further. I am curious whether it is possible to connect something like this to a stream and pull chunks from the stream or let the stream push them. I'll investigate this later with a small example app.

**Comment by inf3rno** [comment]
It is hard to get multiple wrappers from a single statement.

```js
df(a).link(b)
```

This should return a graph or selection instead of a single node. A selection would be better I think. So `df(a)` means we created a new selection and `a` is a member. Then `.link(b)` means that we added a new node and a new link to the selection. The first item of the selection will be `a` and the last node will be `b`. By linking further always the last node will be the source. What if we link multiple nodes e.g. `.link(b,c,d)`. Hmm then all of them will be source by the next link. Hmm by linking we should be able to prev, next, first, last. If the interface would support these, then we would be able to build a complex graph in a single line.

**Comment by inf3rno** [comment]
How would I use this in the practice?

```js
var db = df.node(function (done){
    MongoClient.connect("mongodb://localhost:27017/exampleDb", done)
})
var inputName = df.node(function (done){
    done(null, "inf3rno");
})
var findUsersWithName = df.node(function (done, db, name){
    var users = db.collection('users');
    users.find({name: name}).toArray(done);
})

df.link([db, inputName], findUsersWithName)
findUsersWithName(function (err, usersFound){
    console.log(usersFound)
})
```

This is just a very basic example, but is appears to be more convenient to pull database and input data then trying to inject it somehow. This is similar to `require`, but that does not work by async requirements afaik.

**Comment by inf3rno** [comment]
Is it possible to do the same with objects?

```js
var UserRepository = df.Component.extend({
    prototype: {
        init: function (){
            this.link([this.db, this.inputName], this.findUsersWithName)
        },
        db: function (done){
            MongoClient.connect("mongodb://localhost:27017/exampleDb", done)
        },
        inputName: function (done){
            done(null, "inf3rno");
        },
        findUsersWithName: function (done, db, inputName){
            var users = db.collection('users');
            users.find({name: inputName}).toArray(done);
        }
    }
});

var userRepository = new UserRepository();
userRepository .findUsersWithName(function (err, usersFound){
    console.log(usersFound)
});
```

It is possible, but this way it is too implicit. I need to find out a more explicit solution for objects.

**Comment by inf3rno** [comment]
```js
var UserRepository = df.Component.extend({
    prototype: {
        init: function (){
            this.linkToArgument(this.db, this.findUsersWithName, {arguments: [{index: 0, type: Object}]})
            this.linkToArgument(this.inputName, this.findUsersWithName, {arguments: [{index: 1, type: String}]})
        },
        db: function (){
            MongoClient.connect("mongodb://localhost:27017/exampleDb", this.newCallback())
        },
        inputName: function (){
            this.return("inf3rno");
        },
        findUsersWithName: function (db, inputName){
            var users = db.collection('users');
            users.find({name: inputName}).toArray(this.newCallback());
        }
    }
});

var userRepository = new UserRepository();
userRepository.findUsersWithName(function (err, usersFound){
    console.log(usersFound)
});
```

This is a little bit more explicit.

**Comment by inf3rno** [comment]
```jsr
readerStream = fs.createReadStream('input.txt');
readerStream.setEncoding('UTF8');
rs = df.node(readerStream)
d = df.node(function (chunk){
    console.log(chunk)
})
b = df.node(function (){
    console.log("begin of stream")
})
e = df.node(function (){
    console.log("end of stream")
})
df.link(rs.data, d)
df.link(rs.begin, b)
df.link(rs.end, e)
d()
```

Interesting approach to pull streams, but what if we pull the end event and where to put the chunk reading loop?

**Comment by inf3rno** [comment]
```js
f = new df.FileStream("input.txt")
fn = new df.AsyncFunction(function (done, chunk){
	console.log(chunk)
	done()
})
fn.pulls(f.data)
fn.notifies(fn)
fn.notify()
```

This can be an interesting approach for looping without writing the loop into the callback. The only problem that there is an if-else statement which is not in this structure. We need to add a verification.

```js
fn.notifies(fn, function (done){
    done(null, !f.atEnd())
})
fn.notifies(end, function (done){
    done(null, f.atEnd())
})
```

A little bit complicated for an if-else statement, but it is good for start. Is it possible to make notifications easier? E.g. add a sync if-else statement and a single notification?

```js
fn.notifies(function (done){
    done(null, f.atEnd()?end:fn)
})
```
so this returns the node/async function instead of a boolean

what if I add event handlers?

```js
fn.on("after:return", function (notify){
    if (f.atEnd())
        end()
    else
        fn()
})
```

I love this better, but it might be harder to follow, and there is no auto complete for strings. And ofc. it is almost the same as using the default stream with event handlers...

How would this look like with an AsyncClass?

```js
f = new df.FileStream("input.txt")
fn = new df.AsyncFunction(function (done, chunk){
    console.log(chunk)
    done()
})
end = new df.AsyncFunction(function (done){
    done()
})
fn.pulls(f.data)
fn.notifies(function (done){
    done(null, f.atEnd()?end:fn)
})
fn.notify()
```

How to add all of these to a class?

**Comment by inf3rno** [comment]
```js
AsyncFilePrinter= df.AsyncClass.extend({
    prototype: {
        init: function (f){
            this.f = f
            this.async(this.print)
            this.print.pulls(this.f.data)
            this.print.notifies(this.print, negate(f.atEnd))
        },
        print: function (done, chunk){
            console.log(chunk)
            done()
        }
    }
})
p = new AsyncFilePrinter(new df.FileStream("input.txt"))
p.print()
```

Interesting approach, I am not sure I want to see code like this. It is not that obvious what happens in the background. Maybe after I got used to it.

**Comment by inf3rno** [comment]
We can simplyfy it

```js
AsyncFilePrinter= df.AsyncClass.extend({
    prototype: {
        init: function (f){
            this.f = f
            this.pipe(this.f.data, this.print)
            this.pipe(this.print, this.next)
        },
        print: function (done, chunk){
            console.log(chunk)
            done()
        },
        next: function (done){
                if (this.f.atEnd())
                    this.pull(this.print);
                done()
            }
        }
    }
})
p = new AsyncFilePrinter(new df.FileStream("input.txt"))
p.print()
```

We could make it even better by making it fully compatible with the stream api instead of using a stream wrapper (ofc. we will use one, but in the background).

```js
        init: function (f){
            this.f = f
            this.flow([this.f, "data"])
                .pipe(this.print)
                .pipe(this.next)
        },
```
```js
f = new AsyncFilePrinter(fs.createReadStream("input.txt"))
```

**Comment by inf3rno** [comment]
So how should this piping work?

```js
flow1.pipe(flow2)
```

If flow1 can push and flow2 is pushable, then it listens to the flow1 push events and pushes data to flow2.
If flow2 can pull and flow1 is pullable, then waits for the pulls of the flow2 and notifies flow1 about the pulls, so it can send back data.

Note that `pull -> notify -> return(data)` is the proper way to handle pulls instead of `pull -> notify -> push(data)`. Just as in the current api, push can be paired with wait/read and pull can be paired with sustain/write. I'll use this api, so I need to work on the details now. It must be fully stream compatible without loosing simplicity.

**Comment by inf3rno** [comment]
```js
            this.flow([this.f, "data"])
                .pipe(this.print)
                .pipe(this.next)
```

this means something like

```js
            df.flow([this.f, "data"])
                .pipe(df.flow([this, "print"]))
                .pipe(df.flow([this, "next"]))
```

This is still simple, especially if we remove the df.flow calls with sugar syntax. Let's stick with this for a while, we can add sugar syntax later again.

So this is a basic tree or graph building api.

```js
flow.pipe(flow).pipe(flow)
```
Currently it is just the tree, but with some modifications we can build graphs with a single line of code. But the point is now that how will this connect a stream to a callback function in a pullable way?

By streams there is a readable and a writable property. A readable can be pulled while a writable can be pushed. But I am not sure these are the same things. Readable can mean, that it can be read, but the flow controls when the data is available, so it can push only. That means it is not pullable. Ofc. a file read stream is not that kind of stuff, but there can be such data sources, by which we cannot control when they send the data, we are just waiting for it. We could call it pull, but in reality that is not what happens:

```js
source.create()

target.wait()
source.push(data)
...
source.push(end)
```

Yes, it is similar to

```js
source.create()

target.pull()
source.sustain(data)
...
source.sustain(end)
```

The difference is in the triggers.
The first one is triggered by the creation and triggered again after the chunk is received or it can wait a few msecs, and don't care whether the target can process the chunks before sending the next.
The second one is triggered by the pulls, so it is controlled by the target, so the target can decide when it is ready to process the next chunk, but it does not care whether the source is ready to send it.

Hmm by streams these features are combined, so they both care whether the other is ready to send or receive the next chunk, at least in ideal cases. We could build a lib, which works the same way by default and add a load balancing lib, if we want something different, for example spawning more targets when all of the existing targets are busy. This can be a strategy in cases when careless pushing or pulling is present. So in many cases probably, since only streams support this behavior, and a lot of other sources or targets doesn't. Hmm sources or sinks maybe. ;-)

Another interesting thing that the current wording of df is not that bad. I might be able to reuse the pump word. I think a pump should represent the loop, which pull or pushes the data.

```
 flow1  -pump-> -pipe-> flow2 # push
 flow1 -pipe-> -pump-> flow2 # pull
```
So it depends on which side of the pipe is the pump whether the upstream flow pushes or the downstream flow pulls, or in other words whether the downstream flow is pushed or the upstream flow is pulled. Hmm interesting, how would this look like in practice? I think it is pretty complicated to add pumps everywhere, but maybe not.

```js
flow.pipe(flow2)
flow.pump() // push
flow2.pump() // pull
```

this is okay, as long as we have 2 flows, but what if we have more?

```js
flow.pipe(flow2).pipe(flow3)
flow.pump() // push - still ok
flow2.pump() // pulling from flow or pushing to flow3? probably both?
```

Does flow2.pump mean here that we need to pull data from flow in order to sustain data push to flow3? I am not sure. This can mean that only a single pump can be in a closed system like this, which I am not sure is true. E.g. if we add a buffer, then it can have a pump, which works a different frequency. I think the idea to create a separate object for flow sustaining loops is interesting, but it requires a lot more thinking.

**Comment by inf3rno** [comment]
Tomorrow has come. :D (after a week)

So the pump is always the loop which drives the flow. Actually this is interesting, since in a complex system only a single pump can be enough.

**Comment by inf3rno** [comment]
A pump is just a little more than a circular graph, so I don't think I really need a dedicated class for it, I mean, an `a.pipe(a)` could solve the problem.

**Comment by inf3rno** [comment]
A subgraph can be created by adding a sink to the upper level graph and adding a source to the lower level graph. Between them a simple echo can be enough. So this is a graph separator, which can have a dedicated class.

**Comment by inf3rno** [comment]
A source should be both pushing and pullable and a sink should be both pulling and pushable with a single line of code. For example a simple for loop can be something like this:

```js
var s = df.node(function (done, i, l){
    if (i<l)
        done(i+1, l)
})
s.pipe(s)
s.pipe(x) // x is some node which can receive messages
```

This is okay, because done can call both push or sustain for pull dependent on the situation. The only thing we need here is something which starts the loop. E.g. `s.push(0, 10)` or `x.pull(cb, 0, 10)`. Can a single pull result in more messages? Can a single push result in more messages. Since the source is piped to itself, it can, but I am not sure whether this should be the default behavior. Another thing, that I need a queue for the pipe, because otherwise the source would get all the indexes from the loop first and only after that would the pipe send them to the receiver. I need some alternation by messages sending to itself and to the receiver, which can be done only by introducing a queue or piping asynchronously, but the latter one is costly.

**Comment by inf3rno** [comment]
Hmm I need to add these sentences to a feature description, so it will be easier to develop the exact api later.

**Comment by inf3rno** [comment]
Stream api:
```
source:
  readable - there is data on the stream we can read
  data - chunk is available
  end - no more data available, buffer is empty
  close - related resource is closed
  pipe(sink)/unpipe(sink)
  read(size)
  pause/resume()
sink:
  pipe/unpipe - after a source was un/piped to the sink
  drain - emitted after the write buffer was full, so could not be written, but it became empty again
  finish - when the writing is ended
  close - related resource is closed
  write(chunk)
  end(chunk)
  cork()/uncork()
```

So it uses a different method to end the stream. Should we do the same or should we raise an exception at the end?

**Comment by inf3rno** [comment]
A flow is different from a stream in at least 2 things:
 - a single flow can connect to many other flows via pipes
 - a flow is reusable, so either it should be stateless or the begin/end, open/close should not be states of it

**Comment by inf3rno** [comment]
What can be common?

 - I think the `drain` and `write() -> false` are okay, so somehow we should be able to indicate that the downstream queue is full and we shall wait before pushing the next message.
 - The read/write wording is okay, and I might be able to separate them from the push/pull, but I am not sure yet. By streams the write pushes and the read pulls, which is interesting. These features are completely separated by zmq and you normally don't mix them, or probably you cannot mix them.

**Comment by inf3rno** [comment]
Let's stick with the loop example.

```js
xxx(function (i, l){
    if (i<l)
        this.resolve(i+1, l)
    else
        this.pause()
})
```

Hmm this is interesting. Pause is not the word I am looking for. It should be something that notifies the node that the message is handled and there is not output. I am not sure whether it is necessary to notify. It would serve only debug purposes if we must call a callback at the end of every message handling. The problem that message handling does not always mean that we are sending an output message to other nodes.

 - So another feature that message handling does not always result in an output regardless of having downstream flows.

I think I should add another constraint:

 - The origin of the message should be unknown, so we cannot filter by upstream flow. This is a need because otherwise we cannot decouple flows. If we need to respond to the upstream flows, then we can give id-s to the messages, and filter the responses by id. The id can be an argument, or we can send DTOs with a message id.

Using pause is in interesting approach. To do that we need to buffer the push and pull events too, so by pause the system will be able to save them and by resume we won't need to resend them.

**Comment by inf3rno** [comment]
```js
var s = df.node(function (done, i, l){
    if (i<l)
        done(null, i+1, l)
})
s.pushes(s)
s.inject(0, 10)
```

The `s.pushes(s)` part is interesting here, it is implicit, but it means that after every single write `s` pushes the message to `s`.

What is really good here, that `push` or `pull` is not part of the actual code. What is not so good that we need an implicit rule here: `auto push by every output message`. We can override this behavior by adding a threshold function.

```js
s.pushes(s, {
    threshold: function (done, queue){
        done(null,  queue.length >= 1)
    }
})`
```

**Comment by inf3rno** [comment]
```js
a.pushes(b).when(function (done, a, b){
    done(null, b.isPushable())
}).otherwise(function (done, a, b){
    a.pause()
})
```
This might work, but I have to simplify it, and it is not clear when the threshold checking code should run. Normally it runs only when a produces output, but in order to resume it must run when b becomes pushable.

**Comment by inf3rno** [comment]
I thought about pumps again. Usually when we are talking about drain, pause, resume, the main problem is load balancing. I think load balancing is a high level problem and we must not mix low level processing code, like message transformation, printing, file reading, etc. with it. It should be on the same level where we connect nodes and describes which node pushes, which pulls, etc. It is interesting that if we manage to merge the common parts of the `on.push` + `pull.cb` and `on.pull` + `push.generator` code, then in theory we will be able to extract the pumping code, which contains the loops which fuels the message sending. This way it will be much easier to load balance the code, since we could monitor whether the pipes are sustained and slow down, pause the pumps, or add new pipes accordingly. This looks like great in theory, but I am not certain whether it is always possible to make a flow both pushing and pullable or both pulling and pushable. I will start the work on the flows with both traits and maybe later I try to work with exceptional cases, maybe I find a solution for them as well.

**Comment by inf3rno** [comment]
Let's start with a constantly pushing and a constantly pulling node.

```js
function a(){
    var i = 0;
    while(true){
        this.write(i);
        i++;
        this.push();
    }
}
```

```js
function b(){
    while(true){
        this.pull(function (){
            var i = this.read();
            console.log(i);
        }.bind(this))
    }
}
```

This is the most basic example. Let's make it recursive so it will be async compatible.

```js
var i = 0;
function a(){
    this.write(i);
    i++;
    this.push(a.bind(this));
}
```

```js
function b(){
    this.pull(function (){
        var i = this.read();
        console.log(i);
        b.call(this);
    }.bind(this))
}
```

Another variation:

```js
var a = new Flow({
    init: function (){
        this.i = 0;
        this.on("pushReceived", this.cb.bind(this))
    },
    cb: function (){
        this.write(this.i);
        this.i++;
        this.push();
    }
});
```

```js
var b = new Flow({
    init: function (){
        this.on("pullSustained", this.cb.bind(this));
    },
    cb: function (){
        var i = this.read();
        console.log(i);
        this.pull();
    }
})
```

We can make them both pushing and pullable and both pulling and pushable this way.

```js
var a = new Flow({
    init: function (){
        this.i = 0;
        this.on("pushReceived", this.cb.bind(this))
        this.on("pulled", this.cb.bind(this));
    },
    cb: function (done){
        this.write(this.i);
        this.i++;
        done(); // done is push by default, but it is sustain by answering pulls
    }
});
```

```js
var b = new Flow({
    init: function (){
        this.on("pullSustained", this.cb.bind(this));
        this.on("pushed", this.cb.bind(this));
    },
    cb: function (done){
        var i = this.read();
        console.log(i);
        done(); // done is pull by default, but it is receive by receiving pushes
    }
})
```

So the event handler should get a callback, which calls the next step in the loop.

**Comment by inf3rno** [comment]
What happens by an average stream, e.g. by file copy?

```js
a = fs.createReadStream(...);
b = fs.createWriteStream(...);
a.pipe(b)
```

This means the following in flow terms:

```js
a = new Flow(...);
b = new Flow(...);
a.pipe(b)
a.pushes()
b.pulls() // only if waiting for drain? so this pull restarts the push behavior by empty buffer
```

**Comment by inf3rno** [comment]
Okay, I have a working code with loops:

```js
var a = new Flow({
    i: 0,
    l: 10,
    data: function (done) {
        if (this.i < this.l) {
            this.write(this.i);
            ++this.i;
        }
        done();
    }
});

var b = new Flow({
    data: function (done) {
        var i = this.read();
        console.log(i);
        done();
    }
});

a.pipe(b);
a.pushes();
a.pushDelivered();
```

The pipe() connects the two flow, the pushes() tells the system that it should create a push loop and the pushDelivered() starts the whole loop. This is almost the same code as above. The Flow class has a medium complexity:

```js
var o3 = require("o3"),
    Queue = require("../util/Queue"),
    Set = require("../util/Set"),
    Pipe = require("./Pipe");

var Flow = o3.Class.extend({
    prototype: {
        init: function (config) {
            this.merge(config);
            this.pipes = new Set();
            this.input = new Queue();
            this.output = new Queue();
        },
        write: function (value) {
            this.output.enqueue(value);
        },
        read: function () {
            return this.input.dequeue();
        },
        pipe: function (destination) {
            var pipe = new Pipe(this, destination);
            this.pipes.add(pipe);
            destination.pipes.add(pipe);
        },
        unpipe: function (flow) {
            var pipe = this.pipes.find(function (pointer) {
                return pointer.destination === flow;
            });
            this.pipes.remove(pipe);
            pipe.destination.pipes.remove(pipe);
        },
        push: function () {
            while (this.output.size()) {
                var message = this.output.dequeue();
                this.pipes.each(function (pipe) {
                    if (pipe.source !== this)
                        return;
                    var destination = pipe.destination;
                    destination.input.enqueue(message);
                }, this);
            }
            this.pipes.each(function (pipe) {
                if (pipe.source !== this)
                    return;
                var destination = pipe.destination;
                destination.pushed(this);
            }, this);
        },
        pushed: function (source) {
            this.data(function () {
                source.pushDelivered();
            });
        },
        pushDelivered: function () {
        },
        pull: function () {
            this.pipes.each(function (pipe) {
                if (pipe.destination !== this)
                    return;
                var source = pipe.source;
                source.pulled(this);
            }, this);
        },
        pulled: function (destination) {
            this.data(function () {
                destination.pullSustained();
            }.bind(this))
        },
        pullSustained: function () {
        },
        data: function (done) {
            done();
        },
        pushes: function () {
            this.pushDelivered = function () {
                this.data(function () {
                    if (this.output.size())
                        this.push();
                }.bind(this));
            }
        },
        pulls: function () {
            this.pullSustained = function () {
                if (this.input.size())
                    this.data(function () {
                        this.pull();
                    }.bind(this))
            };
        }
    }
});

module.exports = Flow;
```

It would become simpler if I'd use a collection lib I guess. Anyways it is almost a god object compared to the pipe, which is very small:

```js
var o3 = require("o3");

var Pipe = o3.Class.extend({
    prototype: {
        init: function (source, destination) {
            this.source = source;
            this.destination = destination;
        },
        destroy: function () {
            this.source.unpipe(this.destination);
        }
    }
});

module.exports = Pipe;
```

We should move some part of the code into the pipes, especially the message moving from source.output to destination.input. But that's okay, this is just a proof of concept code. It works.

**Comment by inf3rno** [comment]
Let's check the api again:

```js
var a = new Flow({
    i: 0,
    l: 10,
    data: function (done) {
        if (this.i < this.l) {
            this.write(this.i);
            ++this.i;
        }
        done();
    }
});

var b = new Flow({
    data: function (done) {
        var i = this.read();
        console.log(i);
        done();
    }
});

a.pipe(b);
a.pushes();
a.pushDelivered();
```

The flow creation is straightforward. If the `done` call is not needed, then I can replace it with return and keep the code sync, but I think it will be needed for example by streams. Only the examples will decide that. I can use factory, for example

```js
var a = df.flow({
    i: 0,
    l: 10,
    data: function (done) {
        if (this.i < this.l) {
            this.write(this.i);
            ++this.i;
        }
        done();
    }
});
```

and maybe custom factories as well, e.g.

```js
df.shortcut("custom", function (fn){
    return new Flow({data: fn});
});

var b = df.custom(function (done) {
    var i = this.read();
    console.log(i);
    done();
});
```

Piping is pumping is far from complete.

```js
a.pipe(b);
a.pushes();
a.pushDelivered();
```

I think it should be much simpler than this. I think the following would be okay:

```js
a.pipe(b);
a.pump();
```

So the piping makes the connection and the pump makes and starts the loop which delivers the messages. This can be okay by straight lines, for example:

```js
a.pipe(b).pipe(c)
b.pump()
// b.pulls(a)
// b.pushes(c)
```

But by branches we have some dark spots here, for example:

```js
a.pipe(b)
a.pipe(c)
c.pump()
```

In this case c pulls the messages from a, but b doesn't. So should b get messages from a, or not? 

I don't know the answer. If things like this do not make sense, then it should throw an error. What about

```js
a.pipe(b)
a.pipe(c)
c.pump()
b.pump()
```

In this case both can get messages from a. If we throw error by the first example, then we should throw error by this as well, since b is not pumping when c is. So need to put the two pumps into a transaction. That's relative easy to do, but it makes the code complicated and hard to follow. I need examples, since now this is just a theoretical problem, and I am not sure whether such systems really exist. If so then the default behavior should support those systems, otherwise the code should throw error.

**Comment by inf3rno** [comment]
Let's check the timer example:

Here is the important part with the old (who knows which version) dataflower:

```js
var timerAdapter = new CustomComponent(function () {
	var timer = new Timer({
		events: {
			started: function () {
				this.ports.started.relay(new Message());
			}.bind(this),
			ticked: function (elapsedTime, length) {
				this.ports.ticked.relay(new Message([elapsedTime, length]));
			}.bind(this),
			cleared: function () {
				this.ports.cleared.relay(new Message());
			}.bind(this),
			ended: function () {
				this.ports.ended.relay(new Message());
			}.bind(this)
		}
	});
	this.ports = {
		changeLength: new InputPort({
			callback: function (message) {
				timer.changeLength(message.data);
			},
			component: this
		}),
		start: new InputPort({
			callback: timer.start,
			context: timer,
			component: this
		}),
		started: new OutputPort({
			component: this
		}),
		stop: new InputPort({
			callback: timer.clear,
			context: timer,
			component: this
		}),
		ticked: new OutputPort({
			component: this
		}),
		cleared: new OutputPort({
			component: this
		}),
		ended: new OutputPort({
			component: this
		})
	};
	this.clear = timer.clear.bind(timer);
});


builder.connectAll(
		timerAdapter.ports.changeLength,
		lengthSelect.ports.selected
);
builder.connectAll(
		timerAdapter.ports.start,
		intervalCheckbox.ports.next,
		startButton.ports.clicked
);
builder.connectAll(
		timerAdapter.ports.started,
		startButton.ports.disable,
		stopButton.ports.enable
);
builder.connectAll(
		timerAdapter.ports.stop,
		stopButton.ports.clicked
);
builder.connectAll(
		timerAdapter.ports.ended,
		beepAudio.ports.replay,
		intervalCheckbox.ports.ended
);
builder.connectAll(
		timerAdapter.ports.cleared,
		startButton.ports.enable,
		stopButton.ports.disable,
		display.ports.clear,
		title.ports.clear
);
builder.connectAll(
		timerAdapter.ports.ticked,
		time2Text.ports.stdin
);
builder.connectAll(
		time2Text.ports.stdout,
		display.ports.text,
		title.ports.text
);
```

So we have a timer object with events, we transform that to a node and we connect that node to view related nodes. How would this look like with the new api?

The event handlers are message sources, so we should transform them into nodes:

```js
df(lengthSelect, "selected").pipe(timer.changeLength);
df(startButton, "clicked").and(intervalCheckbox, "next").pipe(timer.start);
df(timer, "started").pipe(startButton.disable).and(stopButton.enable);
df(stopButton, "clicked").pipe(timer.stop);
df(timer, "ended").and(intervalCheckbox, "ended").pipe(beepAudio.replay);
df(timer, "cleared").pipe(startButton.enable).and(stopButton.disable).and(display.clear).and(title.clear);
df(timer, "ticked").pipe(time2Text).pipe(display.text).and(title.text);
```

So we don't really have to write adapters manually, it is relative easy to decide when to create a source, node or sink adapter from the arguments. Ofc. we can make this complicated:

```js
var ticked = new Source(timer, "ticked");
var time2TextTransformation = new Node(time2Text);
var displayInTextBox = new Sink(display.text);
var displayInTitle = new Sink(title.text);

ticked.pipe(time2TextTransformation);
time2TextTransformation.pipe(displayInTextBox);
time2TextTransformation.pipe(displayInTitle);
```

vs

```js
df(timer, "ticked").pipe(time2Text).pipe(display.text).and(title.text);
```

What about the pump?

Actually in this example we don't need pumps everything reacts on the generated events. In other terms the user pumps the messages manually by hitting a button or selecting a duration. After that the timer's internal async loop serves as pump. Interesting, but in most of the cases we have this kind of pumps.

What about the graph?

It is possible to transform the current code into a graph, but I am not sure whether I really need that feature now. Ofc. later the sub-graph will be needed to hide implementation details and add an interface which can be used by the outside nodes. How to define such an interface?

E.g. the timer could be a sub-graph:

```js
var timerGraph = new Graph({
    changeLength: new Flow(timer.changeLength),
    start: new Flow(timer.start),
    started: new Flow(timer, "started"),
    stop: new Flow(timer.stop),
    ended: new Flow(timer, "ended"),
    cleared: new Flow(timer, "cleared"),
    ticked: new Flow(timer, "ticked")
});

var viewGraph = new Graph({
    lengthChanged: new Flow(lengthSelect, "selected"),
    start: new Flow(startButton, "clicked"),
    stop: new Flow(stopButton, "clicked"),
    startable: new Flow(startButton.enable, stopButton.disable),
    stoppable: new Flow(startButton.disable, stopButton.enable),
    intervalEnded: new Flow(intervalCheckbox, "ended", beepAudio.replay),
    nextInterval: new Flow(intervalCheckbox.next),
    displayNothing: new Flow(display.clear, title.clear),
    displayTime: new Flow(time2Text).pipe(new Flow(display.text, title.text))
});

viewGraph.lengthChanged.pipe(timerGraph.changeLength);
viewGraph.start.pipe(timerGraph.nextInterval, timerGraph.start);
timerGraph.started.pipe(viewGraph.stoppable);
viewGraph.stop.pipe(timerGraph.stop);
timerGraph.stopped.pipe(viewGraph.startable);
timerGraph.ended.pipe(viewGraph.intervalEnded);
timerGraph.cleared.pipe(viewGraph.displayNothing);
timerGraph.ticked.pipe(viewGraph.displayTime);
```

The connections are interesting here, but it would be great to add them to another graph. So if that graph is deleted, then the two subgraphs will be disconnected. Another possible solution to inject one graph into the other.

**Comment by inf3rno** [comment]
Hmm interesting idea, that these connections can act like a graph. I don't think so unless I put the ending points into the graph. So those points will be part of 2 graphs.

Another possible method to make this better is thinking about what we are doing here. We have this many lines long description of how these nodes connect to each other. This is information and to make this simpler we might need to compress this information somehow. So less lines might mean easier understanding. Ofc. this is not always true, but we can try it.

How to make this less?

```js
var timerGraph = new Graph({
    changeLength: new Flow(timer.changeLength),
    start: new Flow(timer.start),
    started: new Flow(timer, "started"),
    stop: new Flow(timer.stop),
    ended: new Flow(timer, "ended"),
    cleared: new Flow(timer, "cleared"),
    ticked: new Flow(timer, "ticked")
});

var viewGraph = new Graph({
    lengthChanged: new Flow(lengthSelect, "selected"),
    start: new Flow(startButton, "clicked"),
    stop: new Flow(stopButton, "clicked"),
    startable: new Flow(startButton.enable, stopButton.disable),
    stoppable: new Flow(startButton.disable, stopButton.enable),
    intervalEnded: new Flow(intervalCheckbox, "ended", beepAudio.replay),
    nextInterval: new Flow(intervalCheckbox.next),
    displayNothing: new Flow(display.clear, title.clear),
    displayTime: new Flow(time2Text).pipe(new Flow(display.text, title.text))
});

viewGraph.lengthChanged.pipe(timerGraph.changeLength);
viewGraph.start.pipe(timerGraph.nextInterval, timerGraph.start);
timerGraph.started.pipe(viewGraph.stoppable);
viewGraph.stop.pipe(timerGraph.stop);
timerGraph.stopped.pipe(viewGraph.startable);
timerGraph.ended.pipe(viewGraph.intervalEnded);
timerGraph.cleared.pipe(viewGraph.displayNothing);
timerGraph.ticked.pipe(viewGraph.displayTime);
```

There are a lot of repeating things here: `timerGraph` and `viewGraph` many times.

```js
var connections = new ConnectionSet({
    @context: [viewGraph, timerGraph],
    connections: [
        lengthChanged: "changeLength",
        nextInterval: "start",
        started: "stoppable",
        stop: "stop", //problem
        stopped: "startable",
        ended: "intervalEnded",
        cleared: "displayNothing",
        ticked: "displayTime",
    ]
})
```

If the node names are easy to understand without the graph name and they are unique, then this might be a solution. It has at least one drawback, we need to use strings instead of node references. Anyways I would call this the jsonld approach. I am pretty sure I won't use this except if we are talking about REST services or something like that, which have "external" interfaces. Currently we are talking about classes and objects, which have "internal" interfaces. Internal means here that we are not talking about inter-class/object communication in the same process and not about RPC, or something like that.

**Comment by inf3rno** [comment]
Something is interesting here, we can describe processes from the continuous parts of these connections. For example

 - length changing process: `timer.changeLength <-> view.lengthChanged`, after that we can start thinking about the modifications: `timer.changeLength -> timer.lengthChanged -> view.changeLength -> view.lengthChanged` + `user.interaction -> view.changeLength -> view.lengthChanged -> timer.changeLength -> timer.lengthChanged`. It is much more complex this way, but this is the proper description of this data binding.

 - starting process: `view.nextIntervalAlert <- view.start -> view.pending -> timer.start -> timer.started -> view.started -> view.stoppable`

etc...

Each one of them describes a feature or at least a scenario. This is interesting, so these routes describe scenarios. So we have a flow control graph, which describes a feature, and we have routes on that graph, which describe a scenario. Each route is directed, so it is relative easy to collect the steps of a scenario if we have the starting point. At least in these examples, maybe different ones are different. Hmm this means that for example we can write something like `v.start/v.pending/t.start/t.started/v.started/v.stoppable` if we want to describe a route. This means we can use routes to describe connections.

```js
graph.starting = [
    new Route(view.start, view.nextIntervalAlert),
    new Route(view.start, [view, "pending"], timer.start, [timer, "started"], [view, "started"], [view, "stoppable"])
]
```

We can combine the 2 approaches:

We can define the `timer.start, [timer, "started"]` subRoute as a node in the graph, since the function results in the event always and the function has input, while the event has output. Interesting approach and does not always apply, but in this case it does.

Another thing that `view.start, [view, "pending"], [view, "started"], [view, "stoppable"]` can be a single node as `view.start, [view, "started"]` and the timer starting node is inside this node, so it is some kind of subGraph or subNode? Anyways it would mean that `viewStarting(viewPending, timerStarting) -> viewStoppable`. This is interesting, so if we use the flattened model, it is less compact and makes less sense than this hierarchical model. Since we are talking about graph construction, it does not really matter which approach we use, but it affects how easily we can find a certain node or a certain connection.

**Comment by inf3rno** [comment]
Another possible solution to copy the method names from model to view. Since they always have to be in sync this is special case.

`timer.start / started <-> view.start / started`

**Comment by inf3rno** [comment]
What if we focus on data processing rather than data delivery, wouldn't it be better? This is a bold statement if we are talking about data delivery, but still, we are trying to design an api here, and easy data processing is more important by apis than data delivery. Imho.

What if we have a class with various methods. We somehow get the argument list for each of the methods. And call the methods only when the arguments are available. If the argument is pullable, then it can be pulled, but if not, then the method waits for a push, or throws exception. This is an interesting approach, and it is inpired by the o3 interface draft.

**Comment by inf3rno** [comment]
```js
var x = new df.Async({
    a: function (x, y){
        var z = x + y;
        return z;
    },
    b: function (z){
        console.log(z);
    }
});
```

The only problem here, we don't know that `a` returns `z`. So we need some kind of pipe which collects the output of `a` and ofc. we need some kind of wrapper which waits for `x` and `y` when `a` is called. What if we don't call `a`? What if `a` is called automatically when the inputs are present?

```js
var q = new df.Async({
    a: function (x, y){
        var z = x + y;
        this.output({z: z});
    },
    b: function (z){
        console.log(z);
    }
});

q.push({x: 1});
q.push({y: 2});
```

Okay, what if we have an async output then we send some kind of future or promise?

```js
var q = new df.Async({
    a: function (x, y){
        var p = new df.Promise();
        setTimeout(function (){
            var z = x + y;
            p.fulfill({z: z});
        }, 1);
        return p;
    },
    b: function (z){
        console.log(z);
    }
});

q.push({x: 1}); // nothing happens
q.push({y: 2}); // a(x, y) is called
// wait 1 ms
// promise is fulfilled with z given
// b(z) is called
// console.log 3
```

I like this, at least this small example. But we need to translate somehow the values and names between nodes, because sooner or later every name will be taken, and we'd need to start namespacing param names which would be burdensome.

Let's try with 2 async nodes.

**Comment by inf3rno** [comment]
```js
var x = new df.Async({
    next: function (i){
        var p = new df.Promise();
        setTimeout(function (){
            p.fulfill({j: i+1});
        }, 1);
        return p;
    }
});
var y = new df.Async({
    next: function (j){
        var p = new df.Promise();
        setTimeout(function (){
            p.fulfill({i: j+1});
        }, 1);
        return p;
    }
});
x.push({i: 0});
```

This is very basic, it would work, but this is not a good way to design nodes. Right now there is a namespace collision. So if we would have a class, for example:

```js
var X = df.Async.extend({
	prototype: {
		next: function (i){
			var p = new df.Promise();
			setTimeout(function (){
				p.fulfill({i: i+1});
			}, 1);
			return p;
		}
	}
});

var x1 = new X();
var x2 = new X();

x1.push({i: 0});
```

What would we expect?

I would expect that x1 loops itself with i and x2 is never triggered. Anyway this is chaotic, we need to somehow pipe these instances and put some constraints on naming.

```js
var x1 = new X();
var x2 = new X();
x1.pipe(x2);
x2.pipe(x1);
x1.inject({i: 0});
```

This can be okay, but still the keys in the two objects share the same namespace. So if we would have an internal method in the X class, which is called by the next(), then it would be called in both instances, because the pipes. We need to specify which values are piped.

```js
var x1 = new X();
var x2 = new X();
x1.pipe({i: x2}); // x2 looks for i in x1's ouput (but not in the input)
x2.pipe({i: x1}); // x1 looks for i in x2's output (but no in the input)
x1.inject({i: 0}); // i is added to the input of x1
```

This is fine, we can define this way which async params are shared and we can communicate via message names. So we don't need a 2 dimensional node/port addressing, which is burdensome if you are programming. Still you need to keep in mind the parameter list, but that's okay. You can do something like

```js
var X = df.Async.extend({
	prototype: {
		expose: {next: true},
		next: function (i){
			var p = new df.Promise();
			setTimeout(function (){
				p.fulfill({i: i+1});
			}, 1);
			return p;
		}
	}
});
```

so you can hide private non-async methods. Another way to add method prefix, which is acceptable. Probably adding an async prefix to the async methods is fine.

```js
var X = df.Async.extend({
	prototype: {
		asyncNext: function (i){
			var p = new df.Promise();
			setTimeout(function (){
				p.fulfill({i: i+1});
			}, 1);
			return p;
		}
	}
});
```

So for example a self loop would be

```js
var x = new X();
x.pipe({i: x});
x.inject({i: 0});
```

I think this code should be in a container, factory or module, since it is about initialization of a graph. So we build the graph structure and inject the first message. The `i` should be added to the params the graph supports. So whenever something pulls in `i` on the graph, then it will know that `i` can be created by the `x` node, since it is piped. I really like this, but ofc. almost every api I designed so far was nice at the beginning and failed by complex graphs, so I need to test this out.

**Comment by inf3rno** [comment]
"async" is a too long word for a prefix. We need to use something shorter which can't be mistyped.

```js
var X = df.Async.extend({
	prototype: {
		asyncNext: function (i){
			var p = new df.Promise();
			setTimeout(function (){
				p.fulfill({i: i+1});
			}, 1);
			return p;
		}
	}
});
```

- asyncNext - too long
- asyNext - nope, sound strange
- asNext - maybe, but then sNext is the sync?
- aNext - funny, "a" and "an" are commonly used, so this is not either
- dfNext - nope
- iNext - interfaces, nope
- yNext - nope
- $ - for jquery
- _ for private methods and underscore
- :S 

I'll have an abstract class without this feature implemented and add async as default. So people will be able to extend the abstract base class if they want a custom prefix or a different async method selector. Hmm probably injecting the selector implementation with a factory would be okay as well.

**Comment by inf3rno** [comment]
This api more or less solves that async output problem, but I think it is still a problem that I cannot pull properly from the function blocks. I don't want to auto pull by empty input. I checked and ES6 yield is supported by most of the recent browsers, so adding something like:

```js
		asyncNext: function *(i){
			i = yield df.pull(i);
			...
		}
```

would solve the non sync input problem. so we could call the functions with the future parameter and a yield would wait out while it gets a value from somewhere. this is fine for example by pulls.

waiting out parallel calls is more problematic, since yield has only one output

```js
		asyncNext: function *(i){
			args = yield df.pull(i, j, k);
			...
		}
```

**Comment by inf3rno** [comment]
Using yield might solve the whole ES5 related callback issue as well.

push: `f(a,b,c)`
pull/await: `yield f(a,b,c)`

so transactions/message identifiers/stacks are maintained by js and not by the developer, which makes things easier. if we want to mix the results of multiple calls, then we can always add a higher level buffer/queue/array, which collects them.

this might work and it is easy to do and supported by most of the current environments, so I think I'll no longer care about the ES5 compatibility and do it the ES6 way.

**Comment by inf3rno** [comment]
Ok. ES6 classes are supported by most of the browsers...

**Comment by inf3rno** [comment]
The only way to use async code with yield is yielding something callable, for example a nodejs style function or a promise. So a scheduler can call these and resume the generator when it gets a result. So the generator itself is not async, but the caller of the generator can work in an asnyc way, which is cool. Not a perfect solution for async functions, but much better then breaking up the tasks into many callbacks. Still we have callbacks here by the async part, but it is much better because with generator we can flatten the async code. I think that's all we needed...

**Comment by inf3rno** [comment]
ff, chrome, opera, safari  and node supports async functions, while edge doesn't
http://kangax.github.io/compat-table/es2016plus/#test-async_functions

**Comment by inf3rno** [comment]
Async functions:

```js
async function asf(a){
    var b = new Promise(function (resolve, reject) {
        setTimeout(function(){
            resolve(2);
        }, 100);
    });
    var c = await b;
    return c;
}

async function run(){
    var c = await asf(1);
    console.log(c);
}

run();
```

vs generators:

```js
function run(gen, cb){
    var param;
    var callNext = function (param){
        var step = gen.next(param);
        if (step.done)
            cb(step.value);
        else if (step.value instanceof Function)
            step.value.call(null, function (err, value){
                callNext(value);
            });
        else
            callNext(step.value);
    };
    callNext(param);
}

function* gen(a){
    var b = yield function (cb){
        setTimeout(function (){
            cb(null, a);
        }, 100);
    };
    var c = yield b;
    return c;
}

run(gen(1), console.log);
```

**Comment by inf3rno** [comment]
The short version:

```js
async function asf(a){
    var b = await new Promise(function (resolve, reject) {
        setTimeout(function(){
            resolve(2);
        }, 100);
    });
    return b + 1;
}

asf(1).then(console.log);
```

The interesting thing here, that `b` is a promise, and `b + 1` is converted to a promise too.  Without the `await` keyword this would not be the case... So there is a huge difference between generators and async functions.

**Comment by inf3rno** [comment]
The only problem with generators and async functions, that they are producing promises, which are non-reusable just like streams. I wanted to avoid that by passing only messages instead of futures, but maybe this is a better way, since a future does not need to contain the message id, so we can do the message synchronization with the js engine instead of writing our own code for that...

**Comment by inf3rno** [comment]
In #121 I laid down the basis of the new API. It solves the subgraph problem once and for all. This current thread should describe the graph and node definitions, the graph builders and invent some serializable addreses and typed messages. After that I can start to work on the implementation.
